### PR TITLE
Automatically infer attention_mask when pad_token_id == eos_token_id

### PR DESCRIPTION
**Pull request by tanuj-rai**
# What does this PR do?

This PR updates the _prepare_special_tokens in the generation utilities to automatically infer a correct attention mask when none is provided and pad_token_id == eos_token_id.

Fixes ##41392

## Before submitting
- [ ] This PR fixes a typo or improves the docs (you can dismiss the other checks if that's the case).
- [X] Did you read the [contributor guideline](https://github.com/huggingface/transformers/blob/main/CONTRIBUTING.md#create-a-pull-request),
      Pull Request section?
- [ ] Was this discussed/approved via a Github issue or the [forum](https://discuss.huggingface.co/)? Please add a link
      to it if that's the case.
- [ ] Did you make sure to update the documentation with your changes? Here are the
      [documentation guidelines](https://github.com/huggingface/transformers/tree/main/docs), and
      [here are tips on formatting docstrings](https://github.com/huggingface/transformers/tree/main/docs#writing-source-documentation).
- [ ] Did you write any new necessary tests?

## Who can review?

@ArthurZucker @vasqu

### CODE REVIEW

**Comment by yonikremer** [comment]
Really happy to see a PR request fast!, the code looks good, def better than what I did locally.

**Comment by tanuj-rai** [comment]
> Really happy to see a PR request fast!, the code looks good, def better than what I did locally.

Thanks @yonikremer!

**Review by vasqu — COMMENTED** [review]
Thanks for the PR. I have some thoughts below but cc @gante for generation.

Tbh, I'm not really sure if we want this as tokenizers behave differently
- Not always pad == eos
- Not always are special tokens used (most of the time they should)
- Padding side will influence the logic here, i.e. left padding behaves differently than right padding

This makes it imo hard to make this generalization that we can infer the mask ourselves. The tokenizer should always return the attention mask instead - when do you only pass the input ids or more like how do you preprocess the text?

**Comment by vasqu** [comment]
Sorry, I think I've passed judgement too fast. I see that it only goes into the specific branch of pad == eos so that handles a bit of my issues I had. 

There still is the question about: Are special tokens used (bos/eos)? And do we adjust to the padding side?

**Review by gante — CHANGES_REQUESTED** [review]
(apologies in advance for the assertive tone, I want to make sure the message gets passed across)

Not approving: when PAD = EOS, we don't know the intent behind the token. Maybe it is indeed an intentional sequence of EOS tokens that is encoding some special model-specific property.

This is actually a very old question that has been popping up in the library for the past few years (even before I joined HF). The consensus is that we want to avoid using `_prepare_special_tokens`, which is model-agnostic, and instead let the model-specific tokenizer prepare the correct mask.

As such, instead of adding 30 LOC with error-prone code, this warning can be easily skipped by calling generate with the attention mask, i.e.

```py
model_inputs = tokenizer(...)
model.generate(**model_inputs)
```

👉 always pass the attention mask when possible, it will prevent complex bugs 🤗

**Comment by tanuj-rai** [comment]
> (apologies in advance for the assertive tone, I want to make sure the message gets passed across)
> 
> Not approving: when PAD = EOS, we don't know the intent behind the token. Maybe it is indeed an intentional sequence of EOS tokens that is encoding some special model-specific property.
> 
> This is actually a very old question that has been popping up in the library for the past few years (even before I joined HF). The consensus is that we want to avoid using `_prepare_special_tokens`, which is model-agnostic, and instead let the model-specific tokenizer prepare the correct mask.
> 
> As such, instead of adding 30 LOC with error-prone code, this warning can be easily skipped by calling generate with the attention mask, i.e.
> 
> ```python
> model_inputs = tokenizer(...)
> model.generate(**model_inputs)
> ```
> 
> 👉 always pass the attention mask when possible, it will prevent complex bugs 🤗

Thank you @gante for the detailed explanation — that makes sense.
I understand now why this behavior shouldn’t be handled globally in _prepare_special_tokens.

**Comment by vasqu** [comment]
Closing as discussed here and in the linked issue. Thanks for understanding 🤗
